### PR TITLE
Dungeon Feature: Collected essences

### PR DESCRIPTION
**Pull request by iHDeveloper**
# Info
A feature that provides the ability to track the essences the player collected during the dungeon game.

**Note:** This PR provide fixes to the **Milestone** feature #359.

## Bug Fixes
- (c04e7a4) Check for milestone pattern when the player is not in a dungeon game
- (2916039) **Collected Essences** and **Milestone** data is the same when the player starts a new game.

# Showcases
image

**Edit:** Included Showcases

### CODE REVIEW

File: src/main/java/codes/biscuit/skyblockaddons/core/DungeonPlayer.java [hunk 1]
```diff
@@ -16,17 +16,70 @@
...
-
         public static Milestone zero(DungeonClass dungeonClass) {
             Milestone milestone = new Milestone();
             milestone.setDungeonClass(dungeonClass);
             milestone.setLevel("⓿");
             milestone.setValue("0");
             return milestone;
         }
+
+        private DungeonClass dungeonClass;
```
Comment (ILikePlayingGames) [inline src/main/java/codes/biscuit/skyblockaddons/core/DungeonPlayer.java:27]: Please move these inner classes to the bottom of the file.

File: src/main/java/codes/biscuit/skyblockaddons/features/tabtimers/TabEffectManager.java [hunk 2]
```diff
@@ -21,6 +21,8 @@
  */
 public class TabEffectManager {
 
+    private static final Pattern SERVER_ID_REGEX = Pattern.compile("[0-9]{2}/[0-9]{2}/[0-9]{2}\\s\\s(mini[0-9]{1,3}[A-Za-z])");
```
Comment (ILikePlayingGames) [inline src/main/java/codes/biscuit/skyblockaddons/features/tabtimers/TabEffectManager.java:24]: Please move all of this to its own method in Utils.java, it fits better there. This class is for the potion effects feature only.

File: src/main/java/codes/biscuit/skyblockaddons/core/DungeonPlayer.java (hunk 1 above)
Comment (iHDeveloper) [reply src/main/java/codes/biscuit/skyblockaddons/core/DungeonPlayer.java]: Done! Applied the changes in https://github.com/BiscuitDevelopment/SkyblockAddons/blob/0452a653d5f246a67e98aef61bbaaa57eff79e97/src/main/java/codes/biscuit/skyblockaddons/core/DungeonPlayer.java

File: src/main/java/codes/biscuit/skyblockaddons/features/tabtimers/TabEffectManager.java (hunk 2 above)
Comment (iHDeveloper) [reply src/main/java/codes/biscuit/skyblockaddons/features/tabtimers/TabEffectManager.java]: Done! Applied the change in https://github.com/BiscuitDevelopment/SkyblockAddons/pull/361/commits/ff75803b8217a549df402477bff95f6acc50fa87

It parses the tab list every second. `PlayerListener#onTick` invokes the `Utils#parseTablist` every second.
And, the `TabEffectManager` doesn't have any change in this PR.

**Comment by iHDeveloper** [comment]
I had to force push the review changes in order to maintain a small commits structure. And, to make the merge more smooth to the `development` branch.

**Comment by ILikePlayingGames** [comment]
Thanks for making the changes! When I was testing it, I noticed the display takes up quite a bit of space on screen. Can you please make it so only the essences that were obtained show up (hide the ones that are zeroes)?

**Pull request merged**